components/collator/src/comparison.rs

```diff
@@ -1551,4 +1633,916 @@ impl CollatorBorrowed<'_> {
+                    // weights, which are not generated for alternate=non-ignorable.  There are
+                    // also exactly as many quaternary weights as tertiary weights, so level
+                    // length differences are handled already on tertiary level.  Any
+                    // above-common quaternary weight will compare greater regardless.
+                    quaternaries.append_byte(LEVEL_SEPARATOR_BYTE);
+                } else {
+                    let q = if q == NO_CE_QUATERNARY {
+                        LEVEL_SEPARATOR_BYTE
+                    } else {
+                        (0xfc + ((q >> 6) & 3)) as u8
+                    };
+                    handle_common!(
+                        quaternaries,
+                        q,
+                        common_quaternaries,
+                        QUAT_COMMON,
+                        QUAT_COMMON[WEIGHT_LOW]
+                    );
+                    quaternaries.append_byte(q);
+                }
+            }
+
+            if (non_primary.bits() >> 24) as u8 == LEVEL_SEPARATOR_BYTE {
+                break; // ce == NO_CE
+            }
+        }
+
+        macro_rules! write_level {
+            ($key:ident, $level:expr, $flag:ident) => {
+                if levels & $flag != 0 {
+                    sink.write(state, &[LEVEL_SEPARATOR_BYTE])?;
+                    sink.write(state, &$key.buf)?;
+                }
+            };
+        }
+
+        write_level!(secondaries, Level::Secondary, SECONDARY_LEVEL_FLAG);
+
+        if levels & CASE_LEVEL_FLAG != 0 {
+            sink.write(state, &[LEVEL_SEPARATOR_BYTE])?;
+
+            // Write pairs of nibbles as bytes, except separator bytes as themselves.
+            let mut b = 0;
+            for c in &cases.buf {
+                debug_assert_eq!(*c & 0xf, 0);
+                debug_assert_ne!(*c, 0);
+                if b == 0 {
+                    b = *c;
+                } else {
+                    sink.write_byte(state, b | (*c >> 4))?;
+                    b = 0;
+                }
+            }
+            if b != 0 {
+                sink.write_byte(state, b)?;
+            }
+        }
+
+        write_level!(tertiaries, Level::Tertiary, TERTIARY_LEVEL_FLAG);
+        write_level!(quaternaries, Level::Quaternary, QUATERNARY_LEVEL_FLAG);
+
+        Ok(())
+    }
+}
+
+/// Error indicating that a [`CollationKeySink`] with limited space ran out of space.
+#[derive(Debug, PartialEq, Eq)]
+pub struct TooSmall;
+
+/// A [`std::io::Write`]-like trait for writing to a buffer-like object.
+///
+/// (This crate does not have access to [`std`].)
+pub trait CollationKeySink {
+    /// The type of error the sink may return.
+    type Error;
+
+    /// An intermediate state object used by the sink, which must implement [`Default`].
+    type State;
+
+    /// A result value indicating the final state of the sink (e.g. a number of bytes written).
+    type Output;
```

Comment:
Observation: I'm fine with the three-associated-types solution, although I think it is achievable by implementing the trait on a type that itself contains a slot for intermediate state, so adding the associated types for State and Output doesn't in itself unlock new functionality.